test/job.js

```diff
@@ -0,0 +1,229 @@
+
+    for (let i = 1; i <= 500; i++) {
+      promises.push(insert(req, apos.modules.article, 'article', {}, i));
+    }
+
+    const inserted = await Promise.all(promises);
+    articleIds = inserted.map(doc => doc._id);
+
+    assert(inserted.length === 500);
+    assert(!!inserted[0]._id);
+  });
+
+  let jobTwo;
+  it('can run a batch job', async function () {
+    const req = apos.task.getReq();
+
+    jobTwo = await jobModule.runBatch(
+      req,
+      articleIds,
+      async function(req, id) {
+        await apos.doc.db.updateOne({
+          _id: id
+        }, {
+          $set: {
+            checked: true
+          }
+        });
+      }
+    );
+
+    assert(!!jobTwo.jobId);
+  });
+
+  it('can follow the second job as it works', async function () {
+    const { completed } = await pollJob({
+      route: `${jobModule.action}/${jobTwo.jobId}`
+    }, {
+      jar
+    });
+
+    assert(completed === articleIds.length);
+    const index = Math.floor(Math.random() * (articleIds.length - 1));
+
+    const article = await apos.http.get(`/api/v1/article/${articleIds[index]}`, {
+      jar
+    });
+
+    assert(article.checked === true);
+  });
+
+  const logged = [];
+
+  let jobThree;
+
+  it('can run a generic job', async function () {
+    const req = apos.task.getReq();
+
+    jobThree = await jobModule.run(
+      req,
+      async function(req, reporters) {
+        let count = 1;
+        reporters.setTotal(articleIds.length);
+
+        for (const id of articleIds) {
+          await Promise.delay(3);
+          logged.push(id);
+          if (count % 2) {
+            reporters.success();
+          } else {
+            reporters.failure();
+          }
+          count++;
+        }
+      }
+    );
+
+    assert(!!jobThree.jobId);
+  });
+
+  it('can follow the third job as it works', async function () {
+    const route = `${jobModule.action}/${jobThree.jobId}`;
+    const { total } = await apos.http.get(route, { jar });
+    // Tests setTotal()
+    assert(total === articleIds.length);
+
+    const {
+      completed,
+      good,
+      bad
+    } = await pollJob({
+      route
+    }, {
+      jar
+    });
+
+    assert(completed === articleIds.length);
+    // Tests success()
+    assert(good === (articleIds.length / 2));
+    // Tests failure()
+    assert(bad === (articleIds.length / 2));
+  });
+});
+
+function padInteger (i, places) {
+  let s = i + '';
+  while (s.length < places) {
+    s = '0' + s;
+  }
+  return s;
+}
+
+async function insert (req, pieceModule, title, data, i) {
+  const docData = Object.assign(pieceModule.newInstance(), {
+    title: `${title} #${padInteger(i, 5)}`,
+    slug: `${title}-${padInteger(i, 5)}`,
+    ...data
+  });
+
+  return pieceModule.insert(req, docData);
+};
+
+async function pollJob(job, { jar }) {
+  const {
+    processed,
+    total,
+    good,
+    bad
+  } = await apos.http.get(job.route, { jar });
+
+  if (processed < total) {
+    await new Promise(resolve => {
```

Comment:
This is correct but you already brought in Bluebird, so use `await Promise.delay(msec)`.

Reply:
Or chuck bluebird and put this in a function. You're only using it for Promise.delay (Promise.all is standard).